File: ios/Classes/LevelPlayBannerAdView.m
#import <IronSource/IronSource.h>
#import "LevelPlayBannerAdView.h"
#import "LevelPlayUtils.h"

@interface LevelPlayBannerAdView()<LPMBannerAdViewDelegate>

// MARK: Properties

@property (nonatomic, strong) FlutterMethodChannel *methodChannel;
@property (nonatomic, strong) LPMBannerAdView *bannerAdView;

@end

@implementation LevelPlayBannerAdView

// MARK: Initialization

- (instancetype)initWithFrame:(CGRect)frame
                       viewId:(int64_t)viewId
        levelPlayBinaryMessenger:(id<FlutterBinaryMessenger>)levelPlayBinaryMessenger
        bannerAdView:(LPMBannerAdView *)bannerAdView
        viewType:(NSString *)viewType
{
    self = [super init];
    if ( self )
    {
        self.bannerAdView = bannerAdView;
        [self.bannerAdView setDelegate: self];

        NSString *uniqueChannelName = [NSString stringWithFormat: @"%@_%lld", viewType, viewId];
        self.methodChannel = [FlutterMethodChannel methodChannelWithName: uniqueChannelName binaryMessenger: levelPlayBinaryMessenger];

        __weak typeof(self) weakSelf = self;
        [self.methodChannel setMethodCallHandler:^(FlutterMethodCall *call, FlutterResult result) {
            [weakSelf handleMethodCall:call result:result];
        }];
    }
    return self;
}

// MARK: Method Call Handling

/**
 Handles method calls received from Flutter.

 @param call The method call received from Flutter.
 @param result The result to be sent back to Flutter.
 */
- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSString *methodName = call.method;

    if ([methodName isEqualToString:@"loadAd"]) {
        [self loadAd:result];
    } else if ([methodName isEqualToString:@"destroyBanner"]) {
        [self destroyBanner:result];
    } else if ([methodName isEqualToString:@"pauseAutoRefresh"]) {
        [self pauseAutoRefresh:result];
    } else if ([methodName isEqualToString:@"resumeAutoRefresh"]) {
        [self resumeAutoRefresh:result];
    }  else {
        result(FlutterMethodNotImplemented);
    }
}

// MARK: Banner Ad View Methods

- (void)loadAd:(FlutterResult)result {
    [_bannerAdView loadAdWithViewController: [LevelPlayUtils getRootViewController]];
    result(nil);
}

- (void)destroyBanner:(FlutterResult)result {
    [_bannerAdView destroy];
    result(nil);
}

- (void)pauseAutoRefresh:(FlutterResult)result {
    [_bannerAdView pauseAutoRefresh];
    result(nil);
}

- (void)resumeAutoRefresh:(FlutterResult)result {
    [_bannerAdView resumeAutoRefresh];
    result(nil);
}

// MARK: View

- (LPMBannerAdView *)view
{
    return self.bannerAdView;
}

// MARK: Deallocation

- (void)dealloc
{
    // Destroy the ad
    [self.bannerAdView destroy];
    // Ensure method channel is set
    if (self.methodChannel != nil) {
        // Retain the method channel temporarily to avoid deallocation during the async block
        FlutterMethodChannel *methodChannel = self.methodChannel;
        // Set method call handler to nil on the main thread
        dispatch_async(dispatch_get_main_queue(), ^{
            [methodChannel setMethodCallHandler:nil];
        });
        // Release the retained method channel after async block execution
        self.methodChannel = nil;
    }
}

// MARK: LPMBannerAdViewDelegate

- (void)didLoadAdWithAdInfo:(LPMAdInfo *) adInfo {
    NSDictionary *args = @{
            @"adInfo": [LevelPlayUtils dictionaryForLevelPlayAdInfo:adInfo]
    };
    [LevelPlayUtils invokeMethodOnUiThreadWithChannel: self.methodChannel methodName: @"onAdLoaded" args: args];
}

- (void)didFailToLoadAdWithAdUnitId:(NSString *)adUnitId error:(NSError *)error {
    NSDictionary *args = @{
            @"error": [LevelPlayUtils dictionaryForLevelPlayAdError:error adUnitId:adUnitId]
    };
    [LevelPlayUtils invokeMethodOnUiThreadWithChannel: self.methodChannel methodName: @"onAdLoadFailed" args: args];
}

- (void)didClickAdWithAdInfo:(LPMAdInfo *)adInfo {
    NSDictionary *args = @{
            @"adInfo": [LevelPlayUtils dictionaryForLevelPlayAdInfo:adInfo]
    };
    [LevelPlayUtils invokeMethodOnUiThreadWithChannel: self.methodChannel methodName: @"onAdClicked" args: args];
}

- (void)didDisplayAdWithAdInfo:(LPMAdInfo *)adInfo {
    NSDictionary *args = @{
            @"adInfo": [LevelPlayUtils dictionaryForLevelPlayAdInfo:adInfo]
    };
    [LevelPlayUtils invokeMethodOnUiThreadWithChannel: self.methodChannel methodName: @"onAdDisplayed" args: args];
}

- (void)didFailToDisplayAdWithAdInfo:(LPMAdInfo *)adInfo error:(NSError *)error {
    NSDictionary *args = @{
            @"adInfo": [LevelPlayUtils dictionaryForLevelPlayAdInfo:adInfo],
            @"error": [LevelPlayUtils dictionaryForLevelPlayAdError:error adUnitId:adInfo.adUnitId],
    };
    [LevelPlayUtils invokeMethodOnUiThreadWithChannel: self.methodChannel methodName: @"onAdDisplayFailed" args: args];
}

- (void)didLeaveAppWithAdInfo:(LPMAdInfo *)adInfo {
    NSDictionary *args = @{
            @"adInfo": [LevelPlayUtils dictionaryForLevelPlayAdInfo:adInfo]
    };
    [LevelPlayUtils invokeMethodOnUiThreadWithChannel: self.methodChannel methodName: @"onAdLeftApplication" args: args];
}

- (void)didExpandAdWithAdInfo:(LPMAdInfo *)adInfo {
    NSDictionary *args = @{
            @"adInfo": [LevelPlayUtils dictionaryForLevelPlayAdInfo:adInfo]
    };
    [LevelPlayUtils invokeMethodOnUiThreadWithChannel: self.methodChannel methodName: @"onAdExpanded" args: args];
}

- (void)didCollapseAdWithAdInfo:(LPMAdInfo *)adInfo {
    NSDictionary *args = @{
            @"adInfo": [LevelPlayUtils dictionaryForLevelPlayAdInfo:adInfo]
    };
    [LevelPlayUtils invokeMethodOnUiThreadWithChannel: self.methodChannel methodName: @"onAdCollapsed" args: args];
}

@end

File: android/src/main/kotlin/com/ironSource/ironsource_mediation/LevelPlayBannerAdView.kt
package com.ironSource.ironsource_mediation

import com.ironSource.ironsource_mediation.LevelPlayUtils.Companion.invokeMethodOnUiThread
import com.unity3d.mediation.LevelPlayAdError
import com.unity3d.mediation.LevelPlayAdInfo
import com.unity3d.mediation.banner.LevelPlayBannerAdView
import com.unity3d.mediation.banner.LevelPlayBannerAdViewListener
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

internal class LevelPlayBannerAdView(
    viewId: Int?,
    levelPlayBinaryMessenger: BinaryMessenger,
    viewType: String,
    private var levelPlayBanner: LevelPlayBannerAdView?
) : PlatformView, LevelPlayBannerAdViewListener {
    private var methodChannel: MethodChannel? = null

    init {
        methodChannel = MethodChannel(levelPlayBinaryMessenger, "${viewType}_$viewId")
        methodChannel!!.setMethodCallHandler { call, result ->  handleMethodCall(call, result)}
        levelPlayBanner?.setBannerListener(this)
    }

    /**
     * Handles method calls from Flutter.
     * This method is invoked when a method call is received from Flutter, and it delegates the call to the appropriate handler method.
     *
     * @param call The method call from Flutter.
     * @param result The result to be returned to Flutter.
     */
    private fun handleMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when(call.method) {
            "loadAd" -> loadAd(result)
            "destroyBanner" -> destroyBanner(result)
            "pauseAutoRefresh" -> pauseAutoRefresh(result)
            "resumeAutoRefresh" -> resumeAutoRefresh(result)
            else -> result.error("ERROR", "Method ${call.method} unknown", null)
        }
    }

    private fun loadAd(result: MethodChannel.Result) {
        levelPlayBanner?.loadAd()
        // Return success result to Flutter
        result.success(null)
    }

    private fun destroyBanner(result: MethodChannel.Result) {
        levelPlayBanner?.destroy()
        // Return success result to Flutter
        result.success(null)
    }


    private fun pauseAutoRefresh(result: MethodChannel.Result) {
        levelPlayBanner?.pauseAutoRefresh()
        // Return success result to Flutter
        result.success(null)
    }

    private fun resumeAutoRefresh(result: MethodChannel.Result) {
        levelPlayBanner?.resumeAutoRefresh()
        // Return success result to Flutter
        result.success(null)
    }

    override fun getView(): LevelPlayBannerAdView? = levelPlayBanner

    override fun dispose() {
        levelPlayBanner?.destroy()
        // Set method call handler to nil
        methodChannel?.setMethodCallHandler(null)
        // Set method channel to nil
        methodChannel = null
    }

    override fun onAdLoaded(adInfo: LevelPlayAdInfo) {
        val args = hashMapOf("adInfo" to adInfo.toMap())
        invokeMethodOnUiThread(methodChannel!!, "onAdLoaded", args)
    }

    override fun onAdLoadFailed(error: LevelPlayAdError) {
        val args = hashMapOf("error" to error.toMap())
        invokeMethodOnUiThread(methodChannel!!, "onAdLoadFailed", args)
    }

    override fun onAdDisplayed(adInfo: LevelPlayAdInfo) {
        val args = hashMapOf("adInfo" to adInfo.toMap())
        invokeMethodOnUiThread(methodChannel!!, "onAdDisplayed", args)
    }

    override fun onAdDisplayFailed(adInfo: LevelPlayAdInfo, error: LevelPlayAdError) {
        val args = hashMapOf("adInfo" to adInfo.toMap(), "error" to error.toMap())
        invokeMethodOnUiThread(methodChannel!!, "onAdDisplayFailed", args)
    }

    override fun onAdClicked(adInfo: LevelPlayAdInfo) {
        val args = hashMapOf("adInfo" to adInfo.toMap())
        invokeMethodOnUiThread(methodChannel!!, "onAdClicked", args)
    }

    override fun onAdExpanded(adInfo: LevelPlayAdInfo) {
        val args = hashMapOf("adInfo" to adInfo.toMap())
        invokeMethodOnUiThread(methodChannel!!, "onAdExpanded", args)
    }

    override fun onAdCollapsed(adInfo: LevelPlayAdInfo) {
        val args = hashMapOf("adInfo" to adInfo.toMap())
        invokeMethodOnUiThread(methodChannel!!, "onAdCollapsed", args)
    }

    override fun onAdLeftApplication(adInfo: LevelPlayAdInfo) {
        val args = hashMapOf("adInfo" to adInfo.toMap())
        invokeMethodOnUiThread(methodChannel!!, "onAdLeftApplication", args)
    }
}